### PR TITLE
mmctl: user preferences

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
Extend `mmctl` with support for user preferences:

##### `list`
```sh
$ mmctl user preference list test@example.com
user_id: soa5qdiwnig8dkmekej37p8gmw
category: crt_thread_pane_step
name: soa5qdiwnig8dkmekej37p8gmw
value: 999
------------------------------
user_id: soa5qdiwnig8dkmekej37p8gmw
category: crt_tutorial_triggered
name: soa5qdiwnig8dkmekej37p8gmw
value: 999
------------------------------
user_id: soa5qdiwnig8dkmekej37p8gmw
category: display_settings
name: use_military_time
value: false
------------------------------
user_id: soa5qdiwnig8dkmekej37p8gmw
category: start_trial_modal
name: trial_modal_auto_shown
value: true
...
```

##### `list --category`
```sh
$ mmctl user preference list test@example.com --category display_settings
user_id: soa5qdiwnig8dkmekej37p8gmw
category: display_settings
name: message_display
value: clean
------------------------------
user_id: soa5qdiwnig8dkmekej37p8gmw
category: display_settings
name: channel_display_mode
value: full
------------------------------
user_id: soa5qdiwnig8dkmekej37p8gmw
category: display_settings
name: collapsed_reply_threads
value: threads_view
...
```

##### `get`
```sh
$ mmctl user preference get --category display_settings --name collapsed_reply_threads test@example.com
user_id: soa5qdiwnig8dkmekej37p8gmw
category: display_settings
name: collapsed_reply_threads
value: threads_view
```

##### `update` (or `set` as an alias)
```sh
$ mmctl user preference update --category display_settings --name collapsed_reply_threads --value true test@example.com
Preference display_settings collapsed_reply_threads for test@example.com updated successfully

$ mmctl user preference set --category display_settings --name collapsed_reply_threads --value true test@example.com
Preference display_settings collapsed_reply_threads for test@example.com updated successfully
```

##### `delete`
```sh
$ mmctl user preference delete --category display_settings --name collapsed_reply_threads test@example.com
Preference display_settings collapsed_reply_threads for test@example.com deleted successfully
```

All commands supports multiple users, e.g.:
```sh
$ mmctl user preference update --category display_settings --name collapsed_reply_threads --value true test@example.com test2@example.com
Preference display_settings collapsed_reply_threads for test@example.com updated successfully
Preference display_settings collapsed_reply_threads for test2@example.com updated successfully
```

#### Ticket Link
None.

#### Release Note
```release-note
Extend mmctl with support for user preferences
```

### CODE REVIEW

**Review by marianunez — APPROVED** [review]
Thanks for adding this @lieut-data! This will make it much easier now for admins to manage user preferences 🎉

File: server/cmd/mmctl/commands/user.go [hunk 1]
```diff
@@ -336,6 +376,21 @@ Global Flags:
 {{.InheritedFlags.FlagUsages | trimTrailingWhitespaces}}
 `)
 
+	PreferenceListCmd.Flags().StringP("category", "c", "", "The optional category by which to filter")
+	PreferenceGetCmd.Flags().StringP("category", "c", "", "The category of the preference")
+	PreferenceGetCmd.Flags().StringP("name", "n", "", "The name of the preference")
+	_ = PreferenceGetCmd.MarkFlagRequired("name")
```
Comment (mgdelacroix) [inline server/cmd/mmctl/commands/user.go:382]: Shouldn't `category` be required as well?

File: server/cmd/mmctl/commands/user.go [hunk 2]
```diff
@@ -251,6 +252,45 @@ var MigrateAuthCmd = &cobra.Command{
...
 }
 
+var PreferenceCmd = &cobra.Command{
+	Use:     "preference",
+	Aliases: []string{"pref"},
+	Short:   "Manage user preferences",
+}
+
+var PreferenceListCmd = &cobra.Command{
+	Use:     "list [--category category] [users]",
```
Comment (mgdelacroix) [inline server/cmd/mmctl/commands/user.go:262]: I started to add comments to standardize the format of the arguments (brackets for optional, `<>` for positional, etc), but I realise that we're not applying one across the rest of the tool. I'll create a ticket to tackle this outside this PR

**Review by mgdelacroix — APPROVED** [review]
LGTM! :+1:

**Comment by lieut-data** [comment]
@lindalumitchell, this "loosely" fits under our MS Teams initiatives as it impacts related customers. Not sure if you'd be the right person to help decide how to proceed on QA Review here?

File: server/cmd/mmctl/commands/user.go (hunk 1 above)
Comment (lieut-data) [reply server/cmd/mmctl/commands/user.go:383]: Good catch, thank you!

**Comment by lieut-data** [comment]
@DHaussermann, do you think it would be feasible to include this with v9.6 in March? (That would mean merging by February 2nd.)

**Comment by DHaussermann** [comment]
Yes, @lieut-data. I can look at this by next week. That milestone is very safe. 👍

**Comment by lieut-data** [comment]
@DHaussermann, just checking in here to see if merging tomorrow is still viable?

**Comment by DHaussermann** [comment]
Thanks for the ping on this @lieut-data. I was busy with release testing this week but can now circle back to this. To the question of timing I will prioritize this to see if we can get it merged before the 9.6 brach is cut. 

On that note, I took a look here but, I'm having a local deployment issue maybe you can help with. I don't see your change when I try using your new feature. 

Steps:
- Switch to your `mmctl-user-preferences` branch and `git log` shows me your last commit
- Used `go build -o mmctl ./cmd/mmctl` to ensure I am compiling the code from my working directory (at least **I think** this assumption is safe.) 
- Checked for the new flags in the help and it did show. Tried to use one anyway `go run ./cmd/mmctl user list --category` and the flag is not known

I think using `go run ./cmd/mmctl` should force the use of the binary compiled in the working directory.
 
image
 
 Please let me know if you have any thoughts about why I'm not seeing your change. Maybe we could compare the checksum to what I should be seeing? Or I have misunderstood something.

**Comment by lieut-data** [comment]
@DHaussermann note this is a sub command of the existing `user` command -- does that help?

CleanShot 2024-02-02 at 17 40 58

Also, I realized that the feature complete date for v9.6 is Feb 9th, so I think that gives extra time here if I understand correctly :)

**Comment by DHaussermann** [comment]
Hi @lieut-data yes that helps. I see how I was using the command wrong. 

So now _I think_ I'm using it right and I see that `category` is an optional filter but mow it appears like maybe MMCTL is working but something is missing at the end point I'm trying to use. 
image
I can't see the payload getting constructed so I don't know if the team ID or any other parameter is maybe missing.

When this happens server logs show:
```
{"timestamp":"2024-02-05 16:50:31.435 -05:00","level":"debug","msg":"Next run time for scheduler","caller":"jobs/schedulers.go:155","scheduler_name":"post_persistent_notifications","next_runtime":"2024-02-05 16:53:01.435312 -0500 EST m=+2310.921837542"}
{"timestamp":"2024-02-05 16:50:35.280 -05:00","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:172","method":"GET","url":"/api/v4/users/email/dkh-syadmin@mattermost.com","request_id":"f4tgec5axidkznc7yiq88ahfsh","user_id":"","status_code":"200"}
{"timestamp":"2024-02-05 16:50:35.280 -05:00","level":"debug","msg":"not found handler triggered","caller":"web/web.go:63","path":"/api/v4/users/p8gueaozijr3uyxrih94nncbwr/preferences","code":404,"ip":""}
```

Do you have any other thoughts? 
Also to be honest I don't know which preferences it's trying to get. I thought that would be obvious once I made it work :) 
Could there actually be data missing? My server is running the same branch as the MMCTL rterminal.

**Comment by amyblais** [comment]
@lieut-data There are some failing tests here.

**Comment by DHaussermann** [comment]
@lieut-data I forgot that `--local` uses completely different endpoints. 

Yes, with an authenticated user this does work as expected. I have it working now and I'm running through some tests 👍 

Thank you.

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- CRUD tested the feature
- Successfully tested `get`, `set`, `list` and `delete` 
- For `set` ensured new values not in the store can be added and are functional
- Response is intuitive if the command is malformed, user cannot be found etc..
- Updated various data types
- As a precaution, set a value back and forth and ensured target user could see the change in the UI

LGTM!

Thanks for the help on this @lieut-data. Please advise if you'd like me to create task for `--local` support. I'm not too familiar on adding features to MMCTL but I think 0/5 feature parity makes sense here.

**Comment by lieut-data** [comment]
@DHaussermann, I'd love to see `--local` support here. If you're open to tracking as a separate ticket, that would be awesome!

**Pull request merged**